pkgs/development/compilers/vala/setup-hook.sh

```diff
@@ -11,7 +11,9 @@ disable_incompabile_pointer_conversion_warning() {
     # Work around incompatible function pointer conversion errors with clang 16
     # by setting ``-Wno-incompatible-function-pointer-types` in an env hook.
     # See https://gitlab.gnome.org/GNOME/vala/-/issues/1413.
-    NIX_CFLAGS_COMPILE+=" -Wno-incompatible-function-pointer-types"
+    if echo "${NIX_CFLAGS_COMPILE}" | grep -q -F -w -v -- "-Wno-incompatible-function-pointer-types"; then
+        NIX_CFLAGS_COMPILE+=" -Wno-incompatible-function-pointer-types"
+    fi
```

Comment:
Perhaps we could avoid the grep subprocess?

```suggestion
addFlag() {
  for flag in $NIX_CFLAGS_COMPILE; do
    if [[ $flag = $1 ]; then
      return
    fi
  done
  NIX_CFLAGS_COMPILE+=" $1"
}

addFlag -Wno-incompatible-function-pointer-types
```

Reply:
Hmm, this doesn't work well if the flag itself contains spaces though, like ` -isystem <dir>`.
We can get the flag with spaces correctly as a function argument, but then we cannot simply split the existing assembled string on spaces since that will never match. 

Like, how would we detect that `-foo -bar baz -qux` contains `-bar baz`?

We lose this info when we concatenate everything together in a flat string.
A generalisation of the grep invocation is a lot simpler to get right in this case. 

Is it worth trying to do this in pure bash then?

Reply:
We could also save the `grep` with
```suggestion
    if [[ "${NIX_CFLAGS_COMPILE}" != *-Wno-incompatible-function-pointer-types* ]]; then
        NIX_CFLAGS_COMPILE+=" -Wno-incompatible-function-pointer-types"
    fi
```
or similar.

Reply:
I generalised the approach that @rhendric suggested and added it as a function to setup.sh, together with a test case. It seems to work well, thanks!